### PR TITLE
Allow `object_id` as a column name for ActiveRecord

### LINKED ISSUE #50160: ActiveRecord::DangerousAttributeError:        object_id is defined by Active Record
With ActiveRecord Before https://github.com/rails/rails/pull/45883/files
it was possible to have a polymorphic relation `object`, 
implicitly defining `object_type` and `object_id` columns.

### Expected behavior
Able to use `object` as a name for a polymorphic relationship

### Actual behavior
An exception 
```
ActiveRecord::DangerousAttributeError:
       object_id is defined by Active Record. Check to make sure that you don't have an attribute or method with the same name.
```
is raised 

### System configuration
**Rails version**:
updating from 7.0 to 7.1

Can the `object_id` column be removed from the dangerous attributes list?

Having the column with that name was not breaking anything for us for many years, and now we are blocked from upgrading to rails 7.1.

### PR DESCRIPTION
**Pull request by misdoro**
Fixes #50160

### Motivation / Background

The `object_id` name may be used by polymorphic relations where `object` is the best name, like `notification.object`. In that case two columns are created: `object_id` and `object_type`.

This Pull Request has been created because we have a rails application that uses the `object` as the polymorphic relationship name, and it seems to be the best name. It was allowed until rails 7.1 and was not breaking anything.

Using the `object_id` as a column name [was explicitly allowed here](https://github.com/rails/rails/pull/38990) 

### Detail

This Pull Request removes the `object_id` from the list of not allowed column names. 
For the sake of consitency, the comments and tests are updated where relevant to use the `__id__` method instead.

### Additional information

The method `object_id` [was added to the dangerous method list here](https://github.com/rails/rails/pull/45883)

Using the `object_id` as a column name [was previously explicitly allowed here](https://github.com/rails/rails/pull/38990) 

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+*   Allow using `object_id` as a database column name.
+    It was available before rails 7.1 and may be used as a part of polymorphic relationship to `object` where `object` can be any other database record.
+
+    *Mikhail Doronin*
```
Comment (skipkayhil) [inline activerecord/CHANGELOG.md:4]:
```suggestion
    *Mikhail Doronin*

```

**Review by skipkayhil — COMMENTED** [review]
(Thinking out loud, non-blocking to this PR) I wonder if we could create a RuboCop rule to ban `object_id`... are there any downsides to just using `__id__` everywhere?

**Comment by nvasilevski** [comment]
Wonder what others think but perhaps we shouldn't be changing examples that show the ruby object being the same by calling `object_id`

I'm worried that these examples will become less clear since I expect developers to be much more familiar with `Object#object_id` rather than `Object#__id__` 

Also keeping `object_id` in the documentation still fair because even if you have an `object_id` attribute - the value will be the same for the same ruby object, even if the meaning of the value is different

**Comment by byroot** [comment]
I removed the doc changes.

Also removed the CHANGELOG because I'll backport this to 7-1-stable, so the changelog entry should only be there.

**Comment by byroot** [comment]
Backported as 10bb0c0b954e73e7acf57eab9a1c18e097b322b8

**Comment by akhilgkrishnan** [comment]
> I removed the doc changes.
> 
> Also removed the CHANGELOG because I'll backport this to 7-1-stable, so the changelog entry should only be there.

Not sure we need to still consider this here https://github.com/rails/rails/pull/50148

**Comment by chris72205** [comment]
> Backported as [10bb0c0](https://github.com/rails/rails/commit/10bb0c0b954e73e7acf57eab9a1c18e097b322b8)

Is there a timeline/estimate for when this will make it to a Rails release?  7.1.3?  We were affected by the `object_id` change as well so are having to pin `activerecord` to a specific ref right now.

**Comment by davidwessman** [comment]
Same here, tried and tested via 7-1-stable, but cannot deploy dual-boot to Heroku due to slug-size being too large when using the branch vs a tagged release.

**Pull request merged**